internal/sbi/consumer/nrf_service.go

```diff
@@ -0,0 +1,221 @@
+	client, ok := s.nfDiscClients[uri]
+	if ok {
+		defer s.nfDiscMu.RUnlock()
+		return client
+	}
+
+	configuration := Nnrf_NFDiscovery.NewConfiguration()
+	configuration.SetBasePath(uri)
+	client = Nnrf_NFDiscovery.NewAPIClient(configuration)
+
+	s.nfDiscMu.RUnlock()
+	s.nfDiscMu.Lock()
+	defer s.nfDiscMu.Unlock()
+	s.nfDiscClients[uri] = client
+	return client
+}
+
+func (s *nnrfService) SendSearchNFInstances(
+	nrfUri string, targetNfType, requestNfType models.NfType, param Nnrf_NFDiscovery.SearchNFInstancesParamOpts) (
+	*models.SearchResult, error,
+) {
+	// Set client and set url
+	ausfContext := s.consumer.Context()
+
+	client := s.getNFDiscClient(ausfContext.NrfUri)
+
+	ctx, _, err := ausf_context.GetSelf().GetTokenCtx(models.ServiceName_NNRF_DISC, models.NfType_NRF)
+	if err != nil {
+		return nil, err
+	}
+
+	result, res, err := client.NFInstancesStoreApi.SearchNFInstances(ctx, targetNfType, requestNfType, &param)
+	if err != nil {
+		logger.ConsumerLog.Errorf("SearchNFInstances failed: %+v", err)
+	}
+	defer func() {
+		if resCloseErr := res.Body.Close(); resCloseErr != nil {
+			logger.ConsumerLog.Errorf("NFInstancesStoreApi response body cannot close: %+v", resCloseErr)
+		}
+	}()
+	if res != nil && res.StatusCode == http.StatusTemporaryRedirect {
+		return nil, fmt.Errorf("Temporary Redirect For Non NRF Consumer")
+	}
+
+	return &result, nil
+}
+
+func (s *nnrfService) SendDeregisterNFInstance() (problemDetails *models.ProblemDetails, err error) {
+	logger.ConsumerLog.Infof("Send Deregister NFInstance")
+
+	ctx, pd, err := ausf_context.GetSelf().GetTokenCtx(models.ServiceName_NNRF_NFM, models.NfType_NRF)
+	if err != nil {
+		return pd, err
+	}
+
+	ausfContext := s.consumer.Context()
+	client := s.getNFManagementClient(ausfContext.NrfUri)
+
+	var res *http.Response
+
+	res, err = client.NFInstanceIDDocumentApi.DeregisterNFInstance(ctx, ausfContext.NfId)
+	if err == nil {
+		return problemDetails, err
+	} else if res != nil {
+		defer func() {
+			if resCloseErr := res.Body.Close(); resCloseErr != nil {
+				logger.ConsumerLog.Errorf("DeregisterNFInstance response cannot close: %+v", resCloseErr)
+			}
+		}()
+		if res.Status != err.Error() {
+			return problemDetails, err
+		}
+		problem := err.(openapi.GenericOpenAPIError).Model().(models.ProblemDetails)
+		problemDetails = &problem
+	} else {
+		err = openapi.ReportError("server no response")
+	}
+	return problemDetails, err
+}
+
+func (s *nnrfService) RegisterNFInstance(ctx context.Context) (
+	resouceNrfUri string, retrieveNfInstanceID string, err error,
+) {
+	ausfContext := s.consumer.Context()
+
+	client := s.getNFManagementClient(ausfContext.NrfUri)
+	nfProfile, err := s.buildNfProfile(ausfContext)
+	if err != nil {
+		return "", "", errors.Wrap(err, "RegisterNFInstance buildNfProfile()")
+	}
+
+	var nf models.NfProfile
+	var res *http.Response
+	for {
+		nf, res, err = client.NFInstanceIDDocumentApi.RegisterNFInstance(context.TODO(), ausfContext.NfId, nfProfile)
```

Comment:
```suggestion
		nf, res, err = client.NFInstanceIDDocumentApi.RegisterNFInstance(ctx, ausfContext.NfId, nfProfile)
```

Don't use `context.TODO()` in production. Use context from parameter instead.